src/lib/parquet/src/Flow/Parquet/ParquetFile/Data/DataBuilder.php

```diff
@@ -0,0 +1,211 @@
+<?php declare(strict_types=1);
+
+namespace Flow\Parquet\ParquetFile\Data;
+
+use Flow\Dremel\Dremel;
+use Flow\Parquet\Exception\RuntimeException;
+use Flow\Parquet\Option;
+use Flow\Parquet\Options;
+use Flow\Parquet\ParquetFile\Page\ColumnData;
+use Flow\Parquet\ParquetFile\Schema\Column;
+use Flow\Parquet\ParquetFile\Schema\FlatColumn;
+use Flow\Parquet\ParquetFile\Schema\NestedColumn;
+use Flow\Parquet\ParquetFile\Schema\PhysicalType;
+use Psr\Log\LoggerInterface;
+use Psr\Log\NullLogger;
+
+final class DataBuilder
+{
+    public function __construct(
+        private readonly Options $options,
+        private readonly LoggerInterface $logger = new NullLogger()
+    ) {
+    }
+
+    /**
+     * @psalm-suppress MixedArrayOffset
+     * @psalm-suppress MixedAssignment
+     * @psalm-suppress MixedArgumentTypeCoercion
+     */
+    public function build(ColumnData $columnData, Column $column) : \Generator
+    {
+        $dremel = new Dremel($this->logger);
+
+        if (!$column->isList() && $column->isMap()) {
+            throw new RuntimeException('Flat data builder supports only flat column types, not LIST and MAP.');
+        }
+
+        foreach ($dremel->assemble($columnData->repetitions, $columnData->definitions, $columnData->values) as $value) {
+            yield $this->enrichData($value, $column);
+        }
+    }
+
+    /**
+     * @TODO : This should be optimized by moving it into different class, checking column only once, caching it and using dedicated data transformer for values to improve performance
```

Comment:
yeah, it could be converted to the issue, my plan is to get back to this when I finish with a writer as it might impact the builder, that's why I left it like that for now